internal/proxy/proxy_test.go

```diff
@@ -0,0 +1,170 @@
+		"Some options set, environment parent directories are created": {
+			http: "http://example.com:8080", https: "https://example.com:8080", existingDirs: []string{},
+		},
+		"Some options set, environment file is unchanged": {
+			wantUnchangedFiles: []string{envConfigPath},
+			http:               "http://example.com:8080", https: "https://example.com:8080",
+			prevContents: map[string]string{envConfigPath: fmt.Sprintf(`%s
+HTTP_PROXY=http://example.com:8080
+http_proxy=http://example.com:8080
+HTTPS_PROXY=https://example.com:8080
+https_proxy=https://example.com:8080
+`, proxy.ConfHeader)},
+		},
+		"All options set":                              {http: "http://example.com:8080", https: "https://example.com:8080", ftp: "ftp://example.com:8080", socks: "socks://example.com:8080", noProxy: "localhost,127.0.0.1"},
+		"Password is escaped":                          {http: "http://username:p@$$:w0rd@example.com:8080"},
+		"Escaped password is not escaped again":        {http: "http://username:p%40$$%3Aw0rd@example.com:8080"},
+		"Domain username is escaped":                   {http: `http://EXAMPLE\bobsmith:p@$$:w0rd@example.com:8080`},
+		"Domain username without password is escaped":  {http: `http://EXAMPLE\bobsmith@example.com:8080`},
+		"Escaped domain username is not escaped again": {http: `http://EXAMPLE%5Cbobsmith@example.com:8080`},
+
+		"Proxy files not created when ran with dry-run": {dryRun: true},
+
+		// Error cases
+		"Error when environment config file is read-only":    {existingPerms: map[string]os.FileMode{envConfigPath: 0444}, prevContents: map[string]string{envConfigPath: "something"}, wantErr: true},
+		"Error when environment directory is a file":         {existingDirs: []string{"etc/"}, prevContents: map[string]string{filepath.Dir(envConfigPath): "something"}, wantErr: true},
+		"Error when environment directory is not executable": {existingPerms: map[string]os.FileMode{filepath.Dir(envConfigPath): 0444}, wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.existingDirs == nil {
+				tc.existingDirs = []string{filepath.Dir(envConfigPath)}
+			}
+
+			root := t.TempDir()
+			for _, p := range tc.existingDirs {
+				err := os.MkdirAll(filepath.Join(root, p), 0700)
+				require.NoError(t, err, "Setup: Couldn't create %s", p)
+			}
+
+			for p, c := range tc.prevContents {
+				err := os.WriteFile(filepath.Join(root, p), []byte(c), 0600)
+				require.NoError(t, err, "Setup: Couldn't write previous contents to %q", p)
+
+				err = os.Chtimes(filepath.Join(root, p), time.Now().UTC(), initialTime)
+				require.NoError(t, err, "Setup: Couldn't change mtime for %q", p)
+			}
+
+			for file, perms := range tc.existingPerms {
+				testutils.Chmod(t, filepath.Join(root, file), perms)
+			}
+
+			ctx := context.Background()
+			if tc.dryRun {
+				ctx = context.WithValue(ctx, proxy.DryRun, true)
```

Comment:
I would expose DryRun only for tests, wdyt?

Reply:
I have no strong opinion about it. I would prefer to keep dry-run not exposed, but just take whatever path you prefer between the 2 options.